pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionSnapshotCache.java

```diff
@@ -32,32 +34,114 @@
 
         if (log.isDebugEnabled()) {
             log.debug("[{}] Added new replicated-subscription snapshot at {} -- {}", subscription, position,
                     snapshot.getSnapshotId());
         }
+        // Case 1: cache if empty
+        if (snapshots.lastEntry() == null) {
+            snapshots.put(position, snapshotEntry);
+            return;
+        }
 
-        snapshots.put(position, snapshot);
+        // The time difference between the previous position and the earliest cache entry
+        final long timeSinceFirstSnapshot = publishTime - snapshots.firstEntry().getValue().timestamp();
+        // The time difference between the previous position and the lately cache entry
+        final long timeSinceLastSnapshot = publishTime - snapshots.lastEntry().getValue().timestamp();
+        // The time window length of each time slot, used for dynamic adjustment in the snapshot cache.
+        // The larger the time slot, the slower the update.
+        final long timeWindowPerSlot = timeSinceFirstSnapshot / snapshotFrequencyMillis / maxSnapshotToCache;
 
-        // Prune the cache
-        while (snapshots.size() > maxSnapshotToCache) {
-            snapshots.pollFirstEntry();
+        if (position.compareTo(snapshots.firstKey()) < 0) {
+            // Case 2: Reset cursor if position precedes first entry
+            snapshots.clear();
+            snapshots.put(position, snapshotEntry);
+            return;
+        } else if (position.compareTo(snapshots.lastKey()) < 0) {
+            // Case 3: Reset cursor If the position is in the middle, delete the cache after that position
+            while (position.compareTo(snapshots.lastKey()) < 0) {
+                snapshots.pollLastEntry();
+            }
+            snapshots.put(position, snapshotEntry);
+        }
+        // Time-based eviction conditions
+        // timeSinceLastSnapshot < snapshotFrequencyMillis, keep the same frequency
+        // timeSinceLastSnapshot < timeWindowPerSlot, implementing dynamic adjustments
+        if (timeSinceLastSnapshot < snapshotFrequencyMillis || timeSinceLastSnapshot < timeWindowPerSlot) {
+            return;
+        }
+        if (snapshots.size() < maxSnapshotToCache) {
+            // Case 4: Add to cache if not full
+            snapshots.put(position, snapshotEntry);
+        } else {
+            // Case 5: Median-based eviction when cache is full
+            int medianIndex = maxSnapshotToCache / 2;
+            Position positionToRemove = findPositionByIndex(medianIndex);
+            if (positionToRemove != null) {
+                snapshots.remove(positionToRemove);
+            }
+            snapshots.put(position, snapshotEntry);
         }
     }
 
+    /**
+     * Find the Position in NavigableMap according to the target index.
+     */
+    private Position findPositionByIndex(int targetIndex) {
+        Iterator<Map.Entry<Position, SnapshotEntry>> it = snapshots.entrySet().iterator();
```

Comment:
since this doesn't use the value, `snapshots.keySet().iterator()` could be used directly.